File: src/core/exceptions.py
from typing import Dict, Any, Optional


class MCPError(Exception):
    """Base exception class for all MCP-related errors."""
    def __init__(self, message: str, details: Optional[Dict[str, Any]] = None):
        self.message = message
        self.details = details or {}
        super().__init__(message)


class MCPConnectionError(MCPError):
    """Exception raised for connection-related errors."""
    pass


class MCPProtocolError(MCPError):
    """Exception raised for protocol-related errors."""
    pass


class MCPTimeoutError(MCPError):
    """Exception raised when a request times out."""
    pass


class MCPAuthError(MCPError):
    """Exception raised for authentication errors."""
    pass


class MCPExecutionError(MCPError):
    """Exception raised for code execution errors."""
    def __init__(self, message: str, exit_code: int = 1, output: str = "", details: Optional[Dict[str, Any]] = None):
        self.exit_code = exit_code
        self.output = output
        super().__init__(message, details)


class MCPConfigError(MCPError):
    """Exception raised for configuration errors."""
    pass

File: src/core/exceptions.ts
/**
 * Core exception classes for the DeepExec SDK.
 * 
 * This module provides a comprehensive hierarchy of exceptions
 * that can be thrown by the SDK, allowing for precise error handling.
 */

/**
 * Base exception for all MCP-related errors
 */
export class MCPError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'MCPError';
    Object.setPrototypeOf(this, MCPError.prototype);
  }
}

/**
 * Network-level connection failures
 */
export class MCPConnectionError extends MCPError {
  constructor(message: string, public readonly cause?: Error) {
    super(message);
    this.name = 'MCPConnectionError';
    Object.setPrototypeOf(this, MCPConnectionError.prototype);
  }
}

/**
 * Protocol violations or malformed messages
 */
export class MCPProtocolError extends MCPError {
  constructor(message: string, public readonly requestId?: string) {
    super(message);
    this.name = 'MCPProtocolError';
    Object.setPrototypeOf(this, MCPProtocolError.prototype);
  }
}

/**
 * Operation timed out
 */
export class MCPTimeoutError extends MCPError {
  constructor(message: string, public readonly operationId?: string) {
    super(message);
    this.name = 'MCPTimeoutError';
    Object.setPrototypeOf(this, MCPTimeoutError.prototype);
  }
}

/**
 * Authentication/authorization failures
 */
export class MCPAuthError extends MCPError {
  constructor(message: string) {
    super(message);
    this.name = 'MCPAuthError';
    Object.setPrototypeOf(this, MCPAuthError.prototype);
  }
}

/**
 * Runtime execution errors
 */
export class MCPExecutionError extends MCPError {
  /**
   * Creates a new execution error instance
   * 
   * @param message - Error message
   * @param exitCode - Process exit code (if applicable)
   * @param logs - Execution logs (if available)
   */
  constructor(
    message: string, 
    public readonly exitCode?: number, 
    public readonly logs?: string[]
  ) {
    super(message);
    this.name = 'MCPExecutionError';
    Object.setPrototypeOf(this, MCPExecutionError.prototype);
  }
}

/**
 * Configuration errors
 */
export class MCPConfigError extends MCPError {
  constructor(message: string) {
    super(message);
    this.name = 'MCPConfigError';
    Object.setPrototypeOf(this, MCPConfigError.prototype);
  }
}

/**
 * Resource limit exceeded
 */
export class MCPResourceLimitError extends MCPError {
  constructor(message: string, public readonly resourceType?: string) {
    super(message);
    this.name = 'MCPResourceLimitError';
    Object.setPrototypeOf(this, MCPResourceLimitError.prototype);
  }
}
